cli/lsp/language_server.rs

```diff
@@ -99,296 +106,453 @@ pub(crate) struct Inner {
-    self.config.settings.enable
-  }
-
-  /// Searches assets, open documents and external sources for a line_index,
-  /// which might be performed asynchronously, hydrating in memory caches for
-  /// subsequent requests.
-  pub(crate) async fn get_line_index(
-    &mut self,
-    specifier: ModuleSpecifier,
-  ) -> Result<LineIndex, AnyError> {
-    let mark = self.performance.mark("get_line_index");
-    let result = if specifier.scheme() == "asset" {
-      if let Some(asset) = self.get_asset(&specifier).await? {
-        Ok(asset.line_index)
-      } else {
-        Err(anyhow!("asset is missing: {}", specifier))
-      }
-    } else if let Some(line_index) = self.documents.line_index(&specifier) {
-      Ok(line_index)
-    } else if let Some(line_index) = self.sources.get_line_index(&specifier) {
-      Ok(line_index)
-    } else {
-      Err(anyhow!("Unable to find line index for: {}", specifier))
-    };
-    self.performance.measure(mark);
-    result
-  }
-
-  /// Only searches already cached assets and documents for a line index.  If
-  /// the line index cannot be found, `None` is returned.
-  fn get_line_index_sync(
-    &self,
-    specifier: &ModuleSpecifier,
-  ) -> Option<LineIndex> {
-    let mark = self.performance.mark("get_line_index_sync");
-    let maybe_line_index = if specifier.scheme() == "asset" {
-      if let Some(Some(asset)) = self.assets.get(specifier) {
-        Some(asset.line_index.clone())
-      } else {
-        None
-      }
-    } else {
-      let documents = &self.documents;
-      if documents.contains_key(specifier) {
-        documents.line_index(specifier)
-      } else {
-        self.sources.get_line_index(specifier)
-      }
-    };
-    self.performance.measure(mark);
-    maybe_line_index
+  pub async fn state_snapshot(&self) -> StateSnapshot {
+    self.state.lock().unwrap().snapshot()
   }
 
   async fn get_navigation_tree(
-    &mut self,
+    &self,
     specifier: &ModuleSpecifier,
   ) -> Result<tsc::NavigationTree, AnyError> {
     let mark = self.performance.mark("get_navigation_tree");
-    if let Some(navigation_tree) = self.navigation_trees.get(specifier) {
+    let maybe_navigation_trees = {
+      let state = self.state.lock().unwrap();
+      state.navigation_trees.get(specifier).cloned()
+    };
+    if let Some(navigation_tree) = maybe_navigation_trees {
       self.performance.measure(mark);
-      Ok(navigation_tree.clone())
+      Ok(navigation_tree)
     } else {
+      let state_snapshot = self.state_snapshot().await;
       let res = self
         .ts_server
         .request(
-          self.snapshot(),
+          state_snapshot,
           tsc::RequestMethod::GetNavigationTree(specifier.clone()),
         )
         .await?;
       let navigation_tree: tsc::NavigationTree = serde_json::from_value(res)?;
-      self
+      let mut state = self.state.lock().unwrap();
+      state
         .navigation_trees
         .insert(specifier.clone(), navigation_tree.clone());
```

Comment:
That makes sense - I don't think that a big lock that locks for a long time is the correct approach though. The LSP is an asynchronous protocol and we are really shooting ourselves in the foot by making it synchronous. We need to respond to user interaction as fast as possible - without waiting for long running background tasks - otherwise the experience will be bad.

I think the correct approach here is to version fields and only overwrite them if your version is newer than the one in the state - else the response should be considered "stale" and should be discarded, and the request considered cancelled.

We need to carefully review each place where we "copy, compute, update" to see what the desired behaviour is for stale responses. I can pretty much guarantee that in 90% of cases we want to discard the stale result and consider the request cancelled. I will rework the PR to handle these cases more correctly.